### PR TITLE
[WIP] Upgrade to Ogre 1.11.1

### PR DESCRIPTION
**Pull request by AnotherFoxGuy**
Status: building, can load the main menu

Requires RigsOfRods/ror-dependencies#29

### CODE REVIEW

**Comment by tritonas00** [comment]
Is it playable? Can it load vehicles?

**Comment by AnotherFoxGuy** [comment]
No, because a change in the resource manager it doesn't automatically look for materials outside the resource group.
So the game crashes past the main menu

**Comment by tritonas00** [comment]
i' m looking forward to try this because ogre 1.9 crashes a lot with my intel graphics and 1.8 works fine

File: source/main/MainMenu.h [hunk 1]
```diff
@@ -33,7 +33,7 @@
 
 #include <map>
 #include <OgreFrameListener.h>
-#include <OgreWindowEventUtilities.h>
+#include <Bites/OgreWindowEventUtilities.h>
```
Comment (paroj) [inline source/main/MainMenu.h]: this and following: you should not need to add the Component folder prefix here. The include path should include the subdirectory.

File: source/main/gfx/hydrax/Perlin.cpp [hunk 2]
```diff
@@ -389,7 +389,7 @@ namespace Hydrax{namespace Noise
 			PixelBuffer->lock(Ogre::HardwareBuffer::HBL_DISCARD);
 			const Ogre::PixelBox& PixelBox = PixelBuffer->getCurrentLock();
 
-			Data = static_cast<unsigned short*>(PixelBox.data);
```
Comment (paroj) [inline source/main/gfx/hydrax/Perlin.cpp]: rather change to reinterpret_cast. If you looped over uint16 before, looping over uint8 will produce wrong results.

File: source/main/gfx/skyx/VClouds/DataManager.cpp [hunk 3]
```diff
@@ -599,7 +599,7 @@ namespace SkyX { namespace VClouds
 		buffer->lock(Ogre::HardwareBuffer::HBL_DISCARD);
 		const Ogre::PixelBox &pb = buffer->getCurrentLock();
 
-		Ogre::uint32 *pbptr = static_cast<Ogre::uint32*>(pb.data);
```
Comment (paroj) [inline source/main/gfx/skyx/VClouds/DataManager.cpp]: reinterpret_cast

**Comment by paroj** [comment]
> No, because of a change in the resource manager it doesn't automatically look for materials outside the resource group.

you can use OGRE_RESOURCE_MANAGER_STRICT=0 for porting

**Comment by tritonas00** [comment]
PAGEDGEOMETRY CAELUM and RAPIDJSON not supported or they need code changes to work?

**Comment by tritonas00** [comment]
Just tried it. When i try to load simple terrain i get:

> ItemIdentityException: Could not find font cent_black in TextAreaOverlayElement::setFontName at /home/babis/Downloads/rigsofrods-next-build-db-1.10/ogre/src/ogre-1.11.2/Components/Overlay/src/OgreTextAreaOverlayElement.cpp (line 381)

**Comment by paroj** [comment]
> 23:25:10: Error: ScriptCompiler - unexpected token in cent.fontdef(202): 'glyph'. If this is a legacy script you must prepend the type (e.g. font, overlay).

you can ignore these for 1.11.

> 23:25:10: Error: ScriptCompiler - undefined variable in cent.fontdef(115)

this is the new ScriptCompiler choking up on lines like
https://github.com/RigsOfRods/rigs-of-rods/blob/master/resources/OgreCore/cent.fontdef#L100

write `glyph "}" 0..` to explicitly escape it as string

**Comment by AnotherFoxGuy** [comment]
@paroj I've [compiled OGRE with ```OGRE_RESOURCEMANAGER_STRICT=0```](https://github.com/RigsOfRods/ror-dependencies/pull/29/files#diff-af3b638bc2a3e6c650974192a53c7291R235) but it still crashes with ```Can't assign material 'character' to SubEntity of 'Character0_mesh' because this Material does not exist in group 'MeshesRG'. Have you forgotten to define it in a .material script?```
Full log: https://pastebin.com/J3a5PAhx

What would be the easiest way to fix this?

**Comment by paroj** [comment]
while it certainly is something worth investigating, that error message is unlikely the cause for the crash, as Ogre falls back to basewhite in case the specified material is not found.

Do you have a backtrace for the crash?

**Comment by AnotherFoxGuy** [comment]
Indeed, it crashes somewhere else than I expected

It crashes here: https://github.com/RigsOfRods/rigs-of-rods/blob/master/source/main/gfx/EnvironmentMap.cpp#L50
```
Ogre::Texture::setTextureType OgreTexture.h:131
Ogre::TextureManager::createManual OgreTextureManager.cpp:172
Ogre::TextureManager::createManual OgreTextureManager.h:267
RoR::GfxEnvmap::SetupEnvMap EnvironmentMap.cpp:45
SimController::SetupGameplayLoop RoRFrameListener.cpp:2290
main main.cpp:358
__libc_start_main 0x00007fc584aa0830
_start 0x0000000000513649
```

**Comment by paroj** [comment]
this is strange, as the only reason this might happen is TexturePtr being NULL, which by looking at the code can not happen.
Did you clean your local build after setting OGRE_RESOURCEMANAGER_STRICT=0?

The new precompiled headers sometimes get out of sync with the real headers. Alternatively recompile with OGRE_ENABLE_PRECOMPILED_HEADERS=False.

**Comment by Speciesx** [comment]
> d:\rigs-of-rods\build_x86\cmakefiles\cmaketmp\checkincludefile.c(1): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory [D:\rigs-of-rods\build_x86\CMakeFiles\CMakeTmp\cmTC_6770c.vcxproj]

i get this error

**Comment by tritonas00** [comment]
Ok i re-compiled with https://github.com/RigsOfRods/ror-dependencies and i dont have the font crash.
I get same crash as @AnotherFoxGuy's when trying to load simple terrain

**Comment by paroj** [comment]
out of curiosity I fired up gdb.. the problem is two-fold:
1. `ContentManager::resourceCollision` masks the underlying issue by misbehaving (return false)
2. [the real issue is this (second point)](https://github.com/OGRECave/ogre/blob/master/Docs/1.11-Notes.md#breaking-non-api-changes)

creating the environment texture before loading any referencing materials should do the trick.

**Comment by paroj** [comment]
```diff
diff --git a/resources/materials/ror.material b/resources/materials/ror.material
index edad8d87..a157dd44 100644
--- a/resources/materials/ror.material
+++ b/resources/materials/ror.material
@@ -1807,7 +1807,7 @@ material tracks/map/background
 
 			texture_unit
 			{
-				texture MapRttTex1
+				//texture MapRttTex1
 			}
 		}
 	}
diff --git a/source/main/CMakeLists.txt b/source/main/CMakeLists.txt
index b498bbeb..1b03514b 100644
--- a/source/main/CMakeLists.txt
+++ b/source/main/CMakeLists.txt
@@ -629,7 +629,7 @@ add_custom_target(
 )
 
 IF (UNIX)
-    find_file (plugins_cfg plugins.cfg PATHS ${CMAKE_PREFIX_PATH}/share/OGRE)
+    find_file (plugins_cfg plugins.cfg PATHS ${OGRE_CONFIG_DIR})
 
     add_custom_command(
             TARGET ${BINNAME} POST_BUILD
@@ -648,7 +648,7 @@ IF (UNIX)
             FILES_MATCHING PATTERN "*"
     )
     INSTALL(
-            FILES ${TMP_FILE_DIR}/plugins.cfg
+            FILES ${plugins_cfg}
             DESTINATION games/RoR
     )
 ELSEIF (WIN32)
diff --git a/source/main/gfx/EnvironmentMap.cpp b/source/main/gfx/EnvironmentMap.cpp
index 9fa587ea..853756e6 100644
--- a/source/main/gfx/EnvironmentMap.cpp
+++ b/source/main/gfx/EnvironmentMap.cpp
@@ -42,9 +42,7 @@ void RoR::GfxEnvmap::SetupEnvMap()
 {
     m_is_enabled = !BSETTING("Envmapdisable", false); // TODO: Use a GVar!! Postponed until the 'IMGUI' gVar refactor is merged to upstream ~ only_a_ptr, 08/2017
 
-    m_rtt_texture = Ogre::TextureManager::getSingleton().createManual("EnvironmentTexture",
-        Ogre::ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME, Ogre::TEX_TYPE_CUBE_MAP, 256, 256, 0,
-        Ogre::PF_R8G8B8, Ogre::TU_RENDERTARGET);
+    m_rtt_texture = Ogre::TextureManager::getSingleton().getByName("EnvironmentTexture");
 
     for (int face = 0; face < NUM_FACES; face++)
     {
diff --git a/source/main/gui/GUIInputManager.cpp b/source/main/gui/GUIInputManager.cpp
index 7b8f1098..5ab988a0 100644
--- a/source/main/gui/GUIInputManager.cpp
+++ b/source/main/gui/GUIInputManager.cpp
@@ -17,7 +17,7 @@
     You should have received a copy of the GNU General Public License
     along with Rigs of Rods. If not, see <http://www.gnu.org/licenses/>.
 */
-
+#include <MyGUI_Platform.h>
 #if MYGUI_PLATFORM == MYGUI_PLATFORM_WIN32
 #include <Windows.h>
 #endif
diff --git a/source/main/main.cpp b/source/main/main.cpp
index 64d37291..2d032382 100644
--- a/source/main/main.cpp
+++ b/source/main/main.cpp
@@ -186,6 +186,9 @@ int main(int argc, char *argv[])
 
         App::StartOgreSubsystem();
         Ogre::OverlaySystem* overlay_system = new Ogre::OverlaySystem(); //Overlay init
+        Ogre::TextureManager::getSingleton().createManual("EnvironmentTexture",
+                Ogre::ResourceGroupManager::DEFAULT_RESOURCE_GROUP_NAME, Ogre::TEX_TYPE_CUBE_MAP, 256, 256, 0,
+                Ogre::PF_R8G8B8, Ogre::TU_RENDERTARGET);
 
         App::CreateContentManager();
 ```

**Comment by tritonas00** [comment]
^ it works!

**Comment by tritonas00** [comment]
and finally my intel crashes occured with 1.9 are gone!
2018-08-20-032636_1920x1080_scrot

**Pull request closed without merging**